File: go/src/106.从中序与后序遍历序列构造二叉树.go
/*
 * @lc app=leetcode.cn id=106 lang=golang
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 后序遍历最后一个元素是根节点
// 由于不会有重复元素，所以根据根节点的值，在中序遍历中可以得到左子树节点数量
// 然后分成左右两个子树 
func buildTree(inorder []int, postorder []int) *TreeNode {
  size := len(inorder)
  if size == 0 {
    return nil
  }
  root := &TreeNode{ // 后序遍历最后一个元素就是根节点
    Val: postorder[size-1],
  }
  
  i := 0
  for inorder[i] != root.Val {
    i++
  }
  // inorder[:i] 左子树
  // inorder[i+1:] 右子树
  root.Left = buildTree(inorder[:i], postorder[:i])
  root.Right = buildTree(inorder[i+1:], postorder[i:size-1])
  
  return root
}
// @lc code=end

File: rust/src/106.从中序与后序遍历序列构造二叉树.rs
/*
 * @lc app=leetcode.cn id=106 lang=rust
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn build_tree(
        mut inorder: Vec<i32>,
        mut postorder: Vec<i32>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if inorder.len() == 0 {
            return None;
        }

        let mut root = TreeNode {
            val: *postorder.last().unwrap(),
            left: None,
            right: None,
        };

        let index = inorder.iter().position(|&x| x == root.val).unwrap();

        let inorder_right = inorder.split_off(index);
        let postorder_right = postorder.split_off(index);

        root.left = Self::build_tree(inorder, postorder);
        root.right = Self::build_tree(
            inorder_right[1..].to_vec(),
            postorder_right[..postorder_right.len() - 1].to_vec(),
        );

        Some(Rc::new(RefCell::new(root)))
    }
}
// @lc code=end
